src/incentives/AffinePass.sol

```diff
@@ -0,0 +1,182 @@
+    string public baseURI;
+    bytes32 public merkleRoot;
+    mapping(address => bool) public whitelistedBridge;
+    mapping(address => bool) private _hasMintedGuaranteed;
+    mapping(address => uint256) private _minted; // Total number of NFTs minted by each address
+    mapping(address => uint256) private _whitelistMinted; // Number of NFTs minted by a whitelisted wallet
+
+    Counters.Counter private _tokenIdCounter;
+    
+    event WhitelistMerkleRootUpdated(bytes32 indexed merkleRoot);
+
+    modifier onlyBridge() {
+        require(whitelistedBridge[msg.sender], "Only bridge can call");
+        _;
+    }
+
+    constructor( bytes32 _merkleRoot) 
+        ERC721("Affine Pass", "APASS")          
+    {
+        merkleRoot = _merkleRoot;
+        _tokenIdCounter.increment();
+        whitelistedBridge[msg.sender] = true;
+    }
+
+    function setMerkleRoot(bytes32 _merkleRoot) public onlyOwner {
+        merkleRoot = _merkleRoot;
+        emit WhitelistMerkleRootUpdated(merkleRoot);
+    }
+
+    function stopMint() public onlyOwner {
+        saleIsActive = false;
+        whitelistSaleIsActive = false;
+    }
+
+    function hasMintedWhitelist(address _address) public view returns (bool) {
+        return _whitelistMinted[_address] > 0;
+    }
+
+    function hasMinted(address _address) public view returns (bool) {
+        return _minted[_address] > 0;
+    }
+
+    function isWhitelisted(
+        address user,
+        bytes32[] memory proof
+    ) public view returns (bool) {
+        bytes32 node = keccak256(abi.encodePacked(user));
+        return MerkleProof.verify(proof, merkleRoot, node);
+    }
+
+    function toggleWhitelistSale() public onlyOwner {
+        whitelistSaleIsActive = !whitelistSaleIsActive;
+    }
+
+    function togglePublicSale() public onlyOwner {
+        saleIsActive = !saleIsActive;
+        whitelistSaleIsActive = false;
+    }
+
+    function mintReserve(uint256 amount) public onlyOwner {
+        require(totalSupply() + amount <= MAX_SUPPLY, "Exceeds max supply");
+        require(reserveTokens + amount <= MAX_RESERVE_TOKENS, "Exceeds max reserve supply");
+        reserveTokens += amount;
+        for (uint256 i = 0; i < amount; i++) {
+            uint256 tokenId = _tokenIdCounter.current();
+            _tokenIdCounter.increment();
+            _safeMint(owner(), tokenId);
+        }
+    }
+
+    function hasRemainingSupply() public view returns (bool) {
+        uint256 currentSupply = totalSupply() - reserveTokens;
+        return currentSupply < MAXMINTABLESUPPLY;
+    }
+
+    function mintDrop(address[] memory recipients, uint256[] memory quantities) public onlyOwner {
+        require(recipients.length == quantities.length, "Recipients and quantities length mismatch");
+        for (uint256 i = 0; i < recipients.length; i++) {
+            for (uint256 j = 0; j < quantities[i]; j++) {
+                require(totalSupply() + quantities[i] <= MAX_SUPPLY, "Exceeds max supply");
```

Comment:
This might end up using the reserve tokens. Also this check can be done after the first loop.

Reply:
we can't do the reserve token airdrop here since quantity will be 988 (can't be done with one tx)

Reply:
We can exclude the reserve address from the mintdrop process and hit mintRerserve for it instead